prow/spyglass/gcsartifact_fetcher.go

```diff
@@ -0,0 +1,256 @@
+type GCSMarker struct {
+	XMLName xml.Name `xml:"ListBucketResult"`
+	Marker  string   `xml:"NextMarker"`
+}
+
+// Contents is a single entry returned by the GCS XML API
+type Contents struct {
+	Key string
+}
+
+// GCSReq contains the contents of a GCS XML API list response
+type GCSReq struct {
+	XMLName  xml.Name `xml:"ListBucketResult"`
+	Contents []Contents
+}
+
+// names is a helper function for extracting artifact names in parallel from a GCS XML API response
+func names(content []byte, names chan string, src jobSource, wg *sync.WaitGroup) {
+	defer wg.Done()
+	extracted := GCSReq{
+		Contents: []Contents{},
+	}
+	err := xml.Unmarshal(content, &extracted)
+	if err != nil {
+		logrus.WithFields(fieldsForJob(src)).WithError(err).Error("Error unmarshaling artifact names from XML")
+	}
+	for _, c := range extracted.Contents {
+		names <- c.Key
+	}
+}
+
+// tryGCSXMLGet is a helper function that attempts to list the artifacts available for the given job
+// source with the given scheme and number of retries. It returns whatever it was able to get, even if
+// it runs out of retries.
+func tryGCSXMLGet(src jobSource, scheme string, endpoint string, retries int, maxResults int) [][]byte {
+	bucketName, prefix := extractBucketPrefixPair(src.JobPath())
+	bodies := [][]byte{}
+	marker := GCSMarker{}
+	c := http.Client{}
+	if scheme != "https" {
+		tr := &http.Transport{
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+		c = http.Client{Transport: tr}
+		scheme = "http"
+	}
+	retry := 0
+	for retry < retries {
+		params := url.Values{}
+		params.Add("prefix", prefix)
+		if marker.Marker != "" {
+			params.Add("marker", marker.Marker)
+		}
+		req := url.URL{
+			Scheme:   scheme,
+			Host:     endpoint,
+			Path:     bucketName,
+			RawQuery: params.Encode(),
+		}
+		resp, err := c.Get(req.String())
+		if err != nil {
+			logrus.WithField("request", req.String()).WithFields(fieldsForJob(src)).WithError(err).Error("Error in GCS XML API GET request")
+			retry++
+			continue
+		}
+		body, err := ioutil.ReadAll(resp.Body)
+		resp.Body.Close()
+		if err != nil {
+			logrus.WithField("request", req.String()).WithFields(fieldsForJob(src)).WithError(err).Error("Error reading body of GCS XML API response")
+			retry++
+			continue
+		}
+		bodies = append(bodies, body)
+
+		marker = GCSMarker{}
+		err = xml.Unmarshal(body, &marker)
+		if err != nil {
+			logrus.WithField("request", req.String()).WithFields(fieldsForJob(src)).WithError(err).Error("Error unmarshaling body of GCS XML API response")
+		}
+		if marker.Marker == "" {
+			break
+		}
+	}
+	return bodies
+}
+
+// Artifacts lists all artifacts available for the given job source.
+// Uses the GCS XML API because it is ~2x faster than the golang GCS library
+// for large number of artifacts. It should also be S3 compatible according to
+// the GCS api docs.
+func (af *GCSArtifactFetcher) Artifacts(src jobSource) []string {
```

Comment:
For example, this ought to be a `gcsJobSource`. Despite the fact that my future `erickJobSource` implements the `jobSource` interface it is not something that `GCSArtifactFetcher` can work with.